### DISCUSSION TITLE
Encountered too many geometry errors

### DISCUSSION THREAD

**Discussion by YapingQi** [nrc-cnrc/EGSnrc#924] [Questions]
Hi everyone, I build a chamber in egs_chamber package, and run with an error：

```
******** Encountered 1geometry errors (maximum allowed is 0).You can change this limit with the 'geometry error limit' run control input key. ****  then I changed the geometry error limit = 100000, the codes can run, but have so many info like this : 
************** Last 2 geometry steps:
old region=37, position=(0.931625,0.344161,-0.0112392), direction=(0.00965597,0.00256443,0.99995) intended step=0.115387, new region=1, step=0.0582207
old region=1, position=(0.932188,0.34431,0.0469787), direction=(0.00965597,0.00256443,0.99995) intended step=0.066012, new region=1, step=0.066012
EGS_ConeStack::howfar: called from the outside but I find x=(0.932188,0.34431,0.0469787) to be inside
layer=4 distance to planes=14.0133
distance to outer cone=7.62594e-10

******************************
```

can any experts tell me what's wrong with my codes? Thanks in advance!

**Comment by rtownson** [comment]
First, use egs_view to check for any "disappearing" or "speckled" surfaces. Does anything look like you have two surfaces overlapping, so particles might get confused? Are there any unexpected "holes"? What you see in egs_view is the same as what particles see.

Next, in your conestack set all of the outer radii to be equal. This might mean that you want to "pad" the outside of your shape with whatever the surrounding material is. This can greatly reduce the chance for floating point errors.

If you're still having the problem, do a thorough search through your geometries for any surfaces that are touching, and envelopes where the contents touch any surface of the envelope. Check that your source is not defined exactly on a physical surface. Sometimes you just need to offset things by a micrometer to solve the issue.

**Reply by YapingQi** [reply]
Thank you for these suggestions! I've tried to check the geometries in egs_view, and found no problems. But I'm not so sure about the envelope surface touching thing, can you help to check it? Attached is my code file (file format is .txt), thank you in advance.

[PTW30013.txt](https://github.com/nrc-cnrc/EGSnrc/files/9644707/PTW30013.txt)

**Reply by rtownson** [reply]
This was giving a segmentation fault for me that I didn't troubleshoot.. but if you're getting geometry errors I would try changing the first couple layers in the conestack so that the outer radius is constant:
```
        :start layer:
            thickness = 0.0198
            top radii = 0 0.3475
            bottom radii = 0.0335 0.3475
            media = PMMA521ICRU H2O521ICRU
        :stop layer:
        # 2). PMMA & graphite layers, thick=0.127?,top:r_inner=0,r_outer=0.335,r_inner=0.09,r_outer=0.425
        :start layer:
            thickness = 0.0127
            top radii = 0 0.0335 0.3475
            bottom radii = 0.009 0.0425 0.3475
            media = 170C521ICRU PMMA521ICRU H2O521ICRU
        :stop layer:
```

**Reply by YapingQi** [reply]
There is still the geometry errors after changing the outer radius. I tried using a Co-60 source and no errors have been found. Instead, if I use a phase space source, the errors happend. I am confused by the Z position of the source relative to the chamber, is there any way to show the source? like figure in :      file:///D:/EGS/EGSnrc-manuals/pirs898-html/classEGS__TransformedSource.html

**Reply by rtownson** [reply]
If you'd like to post an image, take a screenshot and then drag & drop it into the comment box here.

I would start by turning off variance reduction techniques (just comment out that section), then turning on track scoring (the tracks are wrong if you have variance reduction on):
```
:start ausgab object definition:
  :start ausgab object:
    name = tracks
    library = egs_track_scoring
  :stop ausgab object:
:stop ausgab object definition:
```
After you run a short simulation to generate a few tracks, load them in egs_view. This should help you see where your source is coming from and make sure everything makes sense. I also recommend putting everything in a big air box, so that the source is contained in air realistically.

**Reply by YapingQi** [reply]
I used an egs_phase_space file as the source instead of an IAEA-format ps file, it seems there are no geometry errors.  Thank you for those kind suggestions!

**Reply by YapingQi** [reply]
Btw, do you have any idea for the segmentation fault (core dumped)? I ran with this problem after the egs_chamber simulation.

**Reply by rtownson** [reply]
> Btw, do you have any idea for the segmentation fault (core dumped)? I ran with this problem after the egs_chamber simulation.

Usually this is due to a mistake in the input, like setting the cavity regions to region numbers that don't exist, etc. Does it happen immediately when the simulation starts, in the middle or at the end? Can you simplify the input file to make the segmentation fault go away, and isolate where the problem is coming from?

**Reply by mxxo** [reply]
Hi @YapingQi, sometimes building EGSnrc with Address Sanitizer can help with debugging segfaults. I would run your input file myself but it looks like I'm missing a necessary source library (`libBEAM_Elekta_nrc2.so`).

**To build EGSnrc with Address Sanitizer support**

Assuming your EGSnrc configuration is called `linux`:

* With a text editor, in `HEN_HOUSE/specs/linux.conf` change `FOPT` and `C_FLAGS` to:
```
FOPT = -O2 -mtune=native -g -fsanitize=address -static-libasan
...
C_FLAGS = -O2 -fPIC -g -fsanitize=address -static-libasan
```
* In `HEN_HOUSE/specs/egspp_linux.conf` change `opt` to:
```
opt = -O2 -mtune=native -g -fsanitize=address -static-libasan
```

Then recompile `egs++` and `egs_chamber`: 

```
cd HEN_HOUSE/egs++ && make realclean && make
cd HEN_HOUSE/user_codes/egs_chamber && make clean && make
```

Then, run `egs_chamber` like normal and you should get a crash report when the segfault occurs with a backtrace that we can use to find the issue.

**Reply by YapingQi** [reply]
Thank you so much, Max Orok! I've build Address Sanitizer on my EGSnrc codes as your direction. One more question, any suggestions for me to find the memory leaks problems. I run an script and Address Sanitizer gave me this: 
[output.txt](https://github.com/nrc-cnrc/EGSnrc/files/10102091/output.txt)

**Reply by rtownson** [reply]
Can you share the most recent version of your input file?

**Reply by YapingQi** [reply]
> Can you share the most recent version of your input file?
Thank you, I've solved this problem. Can I ask another question about modeling ion chamber: I am reading the example file named "example1_co60_Pfactors.egsinp", and am confused by the phase space scoring object, which is composed of two volumes:   "inner_phsp_object" and "outer_phsp_object". Why didn't use just one volume, and what are the functions of these two volumes?

**Reply by rtownson** [reply]
The purpose of the `inner_phsp_object` is that is allows for you to score the phase-space with tighter conformity around the actual volume of interest, rather than just on the surface of the water, as defined by `outer_phsp_object`. You could just use `outer_phsp_object`, but then the subsequent calculations would repeatedly transport the particles through the water between `outer_phsp_object` and `inner_phsp_object` each time, so it is less efficient. If you have many calculation geometries (e.g. you are moving something inside the volume contained by `inner_phsp_object` many times, like doing a chamber depth dose), then repeatedly transporting through the same water outside of `inner_phsp_object` is inefficient. Note that this is a somewhat advanced efficiency enhancing technique, and you don't need to understand/do this.

**Reply by YapingQi** [reply]
> Can you share the most recent version of your input file?

I ran a similar chamber file written by myself and Address Sanitizer gave me this report. Can you help with running the input file and help find where the problem is?  
[output30013.txt](https://github.com/nrc-cnrc/EGSnrc/files/10175726/output30013.txt)

**Reply by YapingQi** [reply]
> > Can you share the most recent version of your input file?
> 
> I ran a similar chamber file written by myself and Address Sanitizer gave me this report. Can you help with running the input file and help find where the problem is? [output30013.txt](https://github.com/nrc-cnrc/EGSnrc/files/10175726/output30013.txt) 


The input file : PTW30013simple  
[PTW30013simple.txt](https://github.com/nrc-cnrc/EGSnrc/files/10175743/PTW30013simple.txt)
just change the file type and run with : egs_chamber -i PTW30013simple.egsinp

**Reply by mxxo** [reply]
In the asan report it says where the out-of-bounds read occurs (line 1209 of `egs_chamber.cpp`): 

```
==172251==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x6020000067f4 at pc 0x563fb3779023 bp 0x7fff0c38ecc0 sp 0x7fff0c38ecb0
READ of size 4 at 0x6020000067f4 thread T0
    #0 0x563fb3779022 in EGS_ChamberApplication::initScoring() /root/es/EGSnrc/HEN_HOUSE/user_codes/egs_chamber/egs_chamber.cpp:1209
```

Looking at the file, line 1209 has reads of both `cs_reg[0]` and `cs_reg[1]`:
https://github.com/nrc-cnrc/EGSnrc/blob/a6fc389c6465c949645380976f38a013a639759c/HEN_HOUSE/user_codes/egs_chamber/egs_chamber.cpp#L1209

I don't have an `egs_chamber` setup to confirm this, but this might be a (probably harmless) bug when only one region is given for cross-section enhancement. If that's correct, the line should be rewritten to first check the `cs_reg` vector has two elements before accessing them, as it is an out-of-bounds read if there's only one number as in some of your enhancement settings, e.g. for: 

```
enhance regions = -1   
```

You can try modifying that line in `egs_chamber.cpp`, recompiling with `make clean && make` and then running the input file again to check if this resolves it. 
```cpp
if (do_cse && cs_reg.size() == 2 && cs_reg[0]<0 && cs_reg[1]<0) {
```

**Reply by rtownson** [reply]
Yes @mxxo is right. `enhance regions = -1` is no longer supported (I see it's still in the example). To set a range of the regions to be enhancement regions, provide a start and end region as negative numbers, e.g. `enhance regions = -1 -1000`. Otherwise, just provide the whole list of region numbers.

**Reply by YapingQi** [reply]
@mxxo @rtownson Thank you, guys. These suggestions are helpful!

**Reply by YapingQi** [reply]
> Yes @mxxo is right. `enhance regions = -1` is no longer supported (I see it's still in the example). To set a range of the regions to be enhancement regions, provide a start and end region as negative numbers, e.g. `enhance regions = -1 -1000`. Otherwise, just provide the whole list of region numbers.

Is there any relation between enhancement and adding magnetic field? I found the calculated doses are the same when adding magnetic field. Btw, what areas the magnetic field works for? Attached are two output files with the similar above input file. Thank you in advance.
[zptw30013_0T.txt](https://github.com/nrc-cnrc/EGSnrc/files/10250924/zptw30013_0T.txt)
[zptw30013_1.5T.txt](https://github.com/nrc-cnrc/EGSnrc/files/10250925/zptw30013_1.5T.txt)

**Reply by YapingQi** [reply]
I also used the example input file and got the same results. The only difference between the two inputs is adding a magnetic field parameter in MC transport parameter. Any ideas about this problem? Thank you.
The input *.egsinp
[try0T.txt](https://github.com/nrc-cnrc/EGSnrc/files/10270078/try0T.txt)
[try1T.txt](https://github.com/nrc-cnrc/EGSnrc/files/10270080/try1T.txt)

The outputs:
[try1T.txt](https://github.com/nrc-cnrc/EGSnrc/files/10270096/try1T.txt)
[try0T.txt](https://github.com/nrc-cnrc/EGSnrc/files/10270097/try0T.txt)

**Reply by rtownson** [reply]
In `egs_home/egs_chamber/Makefile` change the line:
```
EGSPP_USER_MACROS = egs_chamber.macros
```
to instead be
```
EGSPP_USER_MACROS = egs_chamber.macros $(EGS_SOURCEDIR)EEMF_macros.mortran
```
and then recompile egs_chamber with `make clean; make` (on linux).
